core/node/node_framework/src/implementations/layers/settlement_layer_data.rs

```diff
@@ -0,0 +1,193 @@
+use anyhow::Context;
+use zksync_config::configs::contracts::{
+    chain::L2Contracts, ecosystem::L1SpecificContracts, SettlementLayerSpecificContracts,
+};
+use zksync_contracts::getters_facet_contract;
+use zksync_contracts_loader::{get_settlement_layer_from_l1, load_settlement_layer_contracts};
+use zksync_eth_client::EthInterface;
+use zksync_gateway_migrator::switch_to_current_settlement_mode;
+use zksync_types::{
+    settlement::SettlementMode, url::SensitiveUrl, Address, L2ChainId, L2_BRIDGEHUB_ADDRESS,
+};
+use zksync_web3_decl::{client::Client, namespaces::ZksNamespaceClient};
+
+use crate::{
+    implementations::resources::{
+        contracts::{
+            L1ChainContractsResource, L1EcosystemContractsResource, L2ContractsResource,
+            SettlementLayerContractsResource,
+        },
+        eth_interface::{EthInterfaceResource, L2InterfaceResource},
+        pools::{MasterPool, PoolResource},
+        settlement_layer::{SettlementModeResource, SlChainIdResource},
+    },
+    wiring_layer::{WiringError, WiringLayer},
+    FromContext, IntoContext,
+};
+
+/// Wiring layer for [`SettlementLayerData`].
+#[derive(Debug)]
+pub struct SettlementLayerData {
+    l2_contracts: L2Contracts,
+    l1_specific_contracts: L1SpecificContracts,
+    // This contracts are required as a fallback
+    l1_sl_specific_contracts: Option<SettlementLayerSpecificContracts>,
+    l2_chain_id: L2ChainId,
+    multicall3: Option<Address>,
+    gateway_rpc_url: Option<SensitiveUrl>,
+}
+
+impl SettlementLayerData {
+    pub fn new(
+        l1_specific_contracts: L1SpecificContracts,
+        l2_contracts: L2Contracts,
+        l2_chain_id: L2ChainId,
+        multicall3: Option<Address>,
+        l1_sl_specific_contracts: Option<SettlementLayerSpecificContracts>,
+        gateway_rpc_url: Option<SensitiveUrl>,
+    ) -> Self {
+        Self {
+            l2_contracts,
+            l1_specific_contracts,
+            l1_sl_specific_contracts,
+            l2_chain_id,
+            multicall3,
+            gateway_rpc_url,
+        }
+    }
+}
+
+#[derive(Debug, FromContext)]
+#[context(crate = crate)]
+pub struct Input {
+    pub eth_client: EthInterfaceResource,
+    pub pool: PoolResource<MasterPool>,
+}
+
+#[derive(Debug, IntoContext)]
+#[context(crate = crate)]
+pub struct Output {
+    initial_settlement_mode: SettlementModeResource,
+    contracts: SettlementLayerContractsResource,
+    l1_ecosystem_contracts: L1EcosystemContractsResource,
+    l1_contracts: L1ChainContractsResource,
+    sl_chain_id: SlChainIdResource,
+    l2_contracts: L2ContractsResource,
+    l2_eth_client: Option<L2InterfaceResource>,
+}
+
+#[async_trait::async_trait]
+impl WiringLayer for SettlementLayerData {
+    type Input = Input;
+    type Output = Output;
+
+    fn layer_name(&self) -> &'static str {
+        "settlement_layer_data"
+    }
+
+    async fn wire(self, input: Self::Input) -> Result<Self::Output, WiringError> {
```

Comment:
I don't quite like that we put so much logic in `wire`, it is supposed to be simple, ideally without side effects, as simple as constructing tasks and resources and outputting them. But if we stick to it then we can't really get the resources we want here, like contract addresses etc. @popzxc @slowli Curious what you think

Reply:
You already have logic for migrating the settlement layer in gateway migrator, no?
I would prefer to simply configure the node for the settlement layer we have no, and then gateway migrator can reboot the node when needed.

Then this method would truly be doing wiring only.

Reply:
i've tried to make it as simple as possible. 
It's still contains some code. especially for en.